### ISSUE TITLE
The Security Engine on the oBIX server

### ISSUE THREAD

**Issue by Qingtao-Cao** [ONEDC/obix#123]
Hooks can be installed in the Device, Watch and History subsystems on the oBIX server to further consult the Security Engine whether the current operation, described by (subject, object, op_code), is permitted. Only when the answer is yes should the operation be executed.

User-space (different clients) can customise their own rules of access controls which is loaded into the oBIX server at start-up and further interpreted by the Security Engine.

Basically the Security Engine implements the mechanism to enforce user-specific access control policy and enables the oBIX server completely neutral of it.

**Comment by anross** [comment]
Alternative point of view: the server responds to all operations and we let auth/access control be handled further up the stack. ie. if an operation reaches the server from clientA, then oBIX assumes clientA is authorized to make that call and responds appropriately.

**Comment by Qingtao-Cao** [comment]
Thanks Andrew!

No matter how we design and implement it, the 3As (Authentication, Authorisation and Access Control) should be properly separated for sake of scalability.

Current on the server side we have been using Apache + Lighttpd + oBIX, where the Apache server sits at an ideal position to Authenticate a request (a client needs to provide credential to prove they are who they claim to be) and if success, assign a unique client ID to a particular client, which is further passed down to lighttpd and oBIX as an environment variable of the current FCGX request.

Whenever a device contract is created on the oBIX server, its descriptor (and relevant persistent file on hard drive) stores its owner ID information. Whenever any other client attempts to access this device, the Security Engine (or, hooks installed in relevant subsystem) will further search through the rules database whether such attempt is allowed. Put it another way, the Security Engine assumes the role of Access control.

Last but not least, the rules database implements the Authorisation, that is, who is allowed to do what kind of operation on whose object, for example, if we would like to set some rules to have a device registered by clientA (thus identified with owner_id = "client_a") writeable by clientB, but readonly for clientC, the relevant rules may look like this:

client_b client_a read | write
client_c client_a read

If the rules database mentions nothing about how client_d would access client_a, clientD's attempt to access a device contract owned by clientA is denied.

This way we can implement a "white-list" access control.

The best part about the 3As separation is that the oBIX server (assumes Access Control role only) is completely neutral to the other two parts. In particular, the Authorisation (rules database) is specific to customer and belongs to high-level policy thus should be separated from the low-level mechanism to enforce it.

Above are my 2 cents. As long as we can do 3As segregation properly we will be doing it the right way :-)

**Comment by Qingtao-Cao** [comment]
If the access control is implemented somewhere top in the stack, that part must know the unique ID of the object (device contract) that a request attempts to access.

Such information is kept in the oBIX server, specifically, the core data structures used in various subsystems. So it would be easier for oBIX server to collect all needed 3-piece information about an access attempt and then consult the access control rules database.

Or, if the needed 3-piece information can be obtained at a high-level component then it can also consult the access control rules database.

**Comment by Qingtao-Cao** [comment]
The SQLite could be manipulated as the "backend" of the Security Engine to store hundreds of thousands of white-list access control rules.

The biggest gain is to spare the effort to develop relevant toolchain in both userspace and kernel space. Take SELinux for example, libsepol and friends are used to compile rules in text file into policy modules and a big chunk of binary image, which is loaded into Linux kernel at start-up when SELinux security server needs to interpret the binary and re-built the whole set of data structure to represent every single bit of syntax of SELinux, before they can be manipulated at run time.

The knowledge of SQL database design is needed to come up with tables for one particular customer (which further contains a number of staff who may access a rack in a datacentre), classes for different assets, permissions of different classes, identifiers of assets, roles, users etc
